common/bounded-executor/src/lib.rs

```diff
@@ -35,13 +35,17 @@ impl BoundedExecutor {
 
     /// Spawn a [`Future`] on the `BoundedExecutor`. This function is async and
     /// will block if the executor is at capacity.
-    pub async fn spawn<F>(&self, f: F)
+    pub async fn spawn<F>(&self, f: F) -> JoinHandle<F::Output>
     where
-        F: Future<Output = ()> + Send + 'static,
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
     {
         let spawn_permit = self.semaphore.acquire().await;
-        let f = f.map(move |_| drop(spawn_permit));
-        self.executor.spawn(f);
+        let f = f.map(move |ret| {
+            drop(spawn_permit);
+            ret
+        });
+        self.executor.spawn(f)
```

Comment:
Why do we need to send the permit to the task, and then promptly drop the permit?

Ah I see, you're dropping the permit once the task has completed. That makes sense.